quesma/model/metrics_aggregations/rate.go

```
@@ -0,0 +1,281 @@
+		currentDaysConst := currentDays
+		for currentDays > 0 {
+			actualDays += util.DaysInMonth(someTime)
+			currentDays -= 30
+			someTime = someTime.AddDate(0, -1, 0)
+		}
+		fix = float64(currentDaysConst) / float64(actualDays)
+	}
+
+	return model.JsonMap{"value": fix * parentVal * query.multiplier}
+}
+
+func (query *Rate) CalcAndSetMultiplier(parentIntervalInMs int64) {
+	query.parentIntervalInMs = parentIntervalInMs
+	if parentIntervalInMs == 0 {
+		logger.ErrorWithCtx(query.ctx).Msgf("parent interval is 0, cannot calculate rate multiplier")
+		return
+	}
+
+	rateInMs := query.unit.ToMilliseconds(query.ctx)
+	// unit month/quarter/year is special, only compatible with month/quarter/year calendar intervals
+	if query.unit == month || query.unit == quarter || query.unit == year {
+		oneMonthInMs := int64(30 * 24 * 60 * 60 * 1000)
+		if parentIntervalInMs < oneMonthInMs {
+			logger.WarnWithCtx(query.ctx).Msgf("parent interval (%d ms) is not compatible with rate unit %s", parentIntervalInMs, query.unit)
+			return
+		}
+		if query.unit == year {
+			rateInMs = 360 * 24 * 60 * 60 * 1000 // round to 360 days, so year/month = 12, year/quarter = 3, as should be
+		}
+	}
+
+	if rateInMs%parentIntervalInMs == 0 {
+		query.multiplier = float64(rateInMs / parentIntervalInMs)
+	} else {
+		query.multiplier = float64(rateInMs) / float64(parentIntervalInMs)
+	}
+}
+
+func (query *Rate) String() string {
+	return fmt.Sprintf("rate(unit: %s)", query.unit)
+}
+
+func (query *Rate) FieldPresent() bool {
+	return query.fieldPresent
+}
+
+func newRateUnit(ctx context.Context, unit string) RateUnit {
+	switch strings.ToLower(unit) {
+	case "second":
+		return second
+	case "minute":
+		return minute
+	case "hour":
+		return hour
+	case "day":
+		return day
+	case "week":
+		return week
+	case "month":
+		return month
+	case "quarter":
+		return quarter
+	case "year":
+		return year
+	default:
+		logger.WarnWithCtxAndThrottling(ctx, "rate", "unit", "invalid rate unit: %s", unit)
```

We can return an error here instead of returning a special `invalid`  value